### PR TITLE
Handle wakeup for all Parser states

### LINKED ISSUE #45: Error Waking up Parser<P, S> FSM
I'm absolutely positive I'm doing this wrong; I have a `rotor_http::client::Persistent<P, S>` machine, which wraps `rotor_http::client::Parser<P, S>`.

If I use a `Notifier` to wake up my fsm while a request is active, I get a panic in the `Parser` and everything blows up. The source of the panic is this line in the `Parser` wrapper:

https://github.com/tailhook/rotor-http/blob/master/src/client/parser.rs#L576

So if I shouldn't be waking up my fsm while it's already busy, how can I avoid calling `wakeup` in code that doesn't know what that machine is up to? I would've thought you could just ignore `wakeup` on a busy machine, because it's already awake?

### PR DESCRIPTION
**Pull request by KodrAus**
Fixes #45 

I've just used the `intent` function like you've done above for other states. Not sure if this is the correct way to handle all other states except `Idle`.

I've done a simple test binary spanking a connection with wakeups and it's handled them as expected instead of panicking. Should I write a test function for this? If so how should I go about it?

### CODE REVIEW

**Comment by tailhook** [comment]
About the fix: I'm not sure this is a good idea. When we are in the middle of sending request the wakeup event should probably be used to add some info for making a request. E.g. if you want to send a file as a part of the request, you might read the part of the file in separate thread and send a wakeup event to the requester, so the `Requester` can write it to the connection buffer.

So I think you should forward event in all the states that have child state machine present.

The test is appreciated. Just pick up a test from `parser.rs` which has a request body. And put a wakeup event somewhere between push_bytes() calls. (And you know, make sure it breaks something on old version)

**Comment by KodrAus** [comment]
Ah I see what you mean, I've reset my solution and added a failing test for wakeups. If you're happy with the shape of that I'll implement the match arm to make it pass.

File: src/client/parser.rs [hunk 1]
```diff
@@ -561,15 +561,20 @@ impl<M, S> Protocol for Parser<M, S>
...
                 maybe_new_request(transport,
                     self.0.wakeup(&Connection {
                         idle: true,
                     }, scope), scope)
-            },
-            // skip the event, will child state machine when idle
-            me@_ => me.intent(self.0, scope)
+            }
+            _ => {
+                unimplemented!();
```
Comment (KodrAus) [inline src/client/parser.rs]: @tailhook I'm not sure how I can get a reference to a `Request` here to pass to the machine's `wakeup`?

It's pretty easy to wake up the `Client` machine, but can't figure out how I should be waking up the `Requester` on `ParserImpl::ReadHeaders` or `ParserImpl::Response`.

Comment (KodrAus) [inline src/client/parser.rs]: Looks like I might have figured it out. There's that `.with` method on `MessageState` that turns it into a `Request`.

File: src/client/parser.rs [hunk 2]
```diff
@@ -572,8 +572,10 @@ impl<M, S> Protocol for Parser<M, S>
                         idle: true,
                     }, scope), scope)
             },
-            ReadHeaders { machine, request, .. } => {
-                let mut req = request.with(transport.output());
+            ReadHeaders { machine, request, is_head } => {
+                let mut req: Request = request.with(transport.output());
+                req.1 = is_head;
```
Comment (KodrAus) [inline src/client/parser.rs:577]: @tailhook If I don't include this line then machines in the `ReadHeaders` state will panic here: https://github.com/tailhook/rotor-http/blob/master/src/client/parser.rs#L348

How should I test this?

Sorry for making such a fuss in this PR, `rotor` is still a bit of a mystery to me 😄

**Comment by KodrAus** [comment]
@tailhook Is this getting closer to the mark or should I leave this for someone with more `rotor` experience to implement?

**Comment by tailhook** [comment]
Yeah. It looks good. I'm not sure if `is_head` is right, need to investigate it a little bit more.

**Comment by tailhook** [comment]
Yeah, sorry for being silent. We have [tk-http](github.com/swindon-rs/tk-http) based on tokio instead of this one. And as you might notice this library is not maintained anymore.

**Comment by KodrAus** [comment]
No problem :) I've just closed a bunch of my old PRs. Thanks for building `rotor`! I was a fan of the API once I got the hang of it.

**Pull request closed without merging**